Mathlib/NumberTheory/SelbergSieve.lean

```diff
@@ -0,0 +1,240 @@
+/-
+Copyright (c) 2024 Arend Mellendijk. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arend Mellendijk
+-/
+import Mathlib.Analysis.Normed.Field.Basic
+import Mathlib.NumberTheory.ArithmeticFunction
+
+/-!
+# The Selberg Sieve
+
+We set up the working assumptions of the Selberg sieve, define the notion of an upper bound sieve
+and show that every upper bound sieve yields an upper bound on the size of the sifted set. We also
+define the Λ² sieve and prove that Λ² sieves are upper bound sieves. We then diagonalise the main
+term of the Λ² sieve.
+
+We mostly follow the treatment outlined by Heath-Brown in the notes to an old graduate course. One
+minor notational difference is that we write $\nu(n)$ in place of $\frac{\omega(n)}{n}$.
+
+## Results
+ * `siftedSum_le_mainSum_errSum_of_UpperBoundSieve` - Every upper bound sieve gives an upper bound
+ on the size of the sifted set in terms of `mainSum` and `errSum`
+
+## Notation
+The `SelbergSieve.Notation` namespace includes common shorthand for the variables included in the
+`SelbergSieve` structure.
+ * `A` for `support`
+ * `𝒜 d` for `multSum d` - the combined weight of the elements of `A` that are divisible by `d`
+ * `P` for `prodPrimes`
+ * `a` for `weights`
+ * `X` for `totalMass`
+ * `ν` for `nu`
+ * `y` for `level`
+ * `R d` for `rem d`
+ * `g d` for `selbergTerms d`
+
+## References
+
+ * [Heath-Brown, *Lectures on sieves*][heathbrown2002lecturessieves]
+ * [Koukoulopoulos, *The Distribution of Prime Numbers*][MR3971232]
+
+-/
+
+noncomputable section
+
+open scoped BigOperators ArithmeticFunction
+
+open Finset Real Nat
+
+/-- We set up a sieve problem as follows. Take a finite set of natural numbers `A`, whose elements
+are weighted by a sequence `a n`. Also take a finite set of primes `P`, represented by a squarefree
+natural number. These are the primes that we will sift from our set `A`. Suppose we can approximate
+`∑ n ∈ {k ∈ A | d ∣ k}, a n = ν d * X + R d`, where `X` is an approximation to the total size of `A`
+and `ν` is a multiplicative arithmetic function such that `0 < ν p < 1` for all primes `p ∣ P`.
+
+Then a sieve-type theorem will give us an upper (or lower) bound on the size of the sifted sum
+`∑ n ∈ {k ∈ support | k.Coprime P}, a n`, obtained by removing any elements of `A` that are a
+multiple of a prime in `P`. -/
+class BoundingSieve where
+  /-- The set of natural numbers that is to be sifted. The fundamental lemma yields an upper bound
+    on the size of this set after the multiples of small primes have been removed. -/
+  support : Finset ℕ
+  /-- The finite set of prime numbers whose multiples are to be sifted from `support`. -/
+  prodPrimes : ℕ
```

Comment:
Could you please provide some motivation why you work with the product of these primes, instead of with the finset of the primes?